pkg/cloud/rgraph/exec/executor_parallel.go

```diff
@@ -104,6 +100,48 @@ func (ex *parallelExecutor) Run(ctx context.Context) (*Result, error) {
 
 }
 
+func (ex *parallelExecutor) deepCopyResult() *Result {
+	ex.lock.Lock()
+	defer ex.lock.Unlock()
+
+	resultCopy := Result{
+		Completed: make([]Action, len(ex.result.Completed)),
+		Pending:   make([]Action, len(ex.result.Pending)),
+		Errors:    make([]ActionWithErr, len(ex.result.Errors)),
+	}
+	copy(resultCopy.Completed, ex.result.Completed)
+	copy(resultCopy.Errors, ex.result.Errors)
+	copy(resultCopy.Pending, ex.result.Pending)
+
+	return &resultCopy
+}
+
+func runWithTimeout(ctx context.Context, timeout time.Duration, runFn func(context.Context) error) error {
+	subctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	return runFn(subctx)
+}
+
+func (ex *parallelExecutor) runActionQueue(ctx context.Context) error {
+	if ex.config.Timeout == 0 {
+		return ex.pq.Run(ctx, ex.runAction)
+	}
+	klog.V(4).Infof("Run runAction with timeout %v.", ex.config.Timeout)
+	f := func(runCtx context.Context) error {
+		return ex.pq.Run(runCtx, ex.runAction)
+	}
+	return runWithTimeout(ctx, ex.config.Timeout, f)
+}
```

Comment:
Please consider removing `runWithTimeout` function above and trying the code below:
```suggestion
func (ex *parallelExecutor) runActionQueue(ctx context.Context) error {
	var cancel context.CancelFunc
	if ex.config.Timeout > 0 {
                ctx, cancel = context.WithTimeout(ctx, ex.config.Timeout)
		defer cancel()
		klog.V(4).Infof("Run runAction with timeout %v.", ex.config.Timeout)
	}
	return ex.pq.Run(ctx, ex.runAction)
}
```

Reply:
done - also with second function (wait for orphans)